api/python/t4/packages.py

```diff
@@ -838,15 +838,22 @@ def set(self, logical_key, entry=None, meta=None):
             self
         """
         if not logical_key or logical_key.endswith('/'):
-            raise QuiltException("Invalid logical_key: %r; cannot be a directory" % logical_key)
+            raise QuiltException(
+                f"Invalid logical key {logical_key!r}. "
+                f"A package entry logical key cannot be a directory."
+            )
+
+        validate_key(logical_key)
 
         if not entry:
             current_working_dir = pathlib.Path.cwd()
             logical_key_abs_path = pathlib.Path(logical_key).absolute()
             entry = logical_key_abs_path.relative_to(current_working_dir)
 
         if isinstance(entry, (string_types, getattr(os, 'PathLike', str))):
-            url = fix_url(str(entry))
+            entry_str = str(entry)
```

Comment:
These are physical keys - shouldn't validate those.

It's totally reasonable to call `pkg.set('foo', '../../foo.txt')`, and `fix_url` will take care of that by resolving relative paths. It won't do anything about proper URLs that contain `/../`... but that's a different problem. Technically, they're legal URLs, and shouldn't break anything. We can test that.

Reply:
The only URLs that we currently accept are of the following form:

```
file://$USERPATH/$FILENAME
s3://$BUCKET/$OBJECT_PATH
```

If `$USERPATH`, `$FILENAME`, or `$OBJECT_PATH` contains a `.` or `..` path fragment that's an error. If `$BUCKET` is `.` or `..`, then you have a bucket named `.` or `..`, which seems impossible?

Reply:
Resolved per in-person discussion.